Automattic/woocommerce-payments
# Failing payment transaction returns to checkout page

Fixes #2469 

#### Changes proposed in this Pull Request


# Failing Transaction Does Not Succeed

In our process redirect payment flow, we currently check to see if we have a valid payment method type (i.e. `card`, `giropay`, or `sofort`) before we check if there are any errors on the payment intent. The problem here is that if there is an error, then the former fields are not present. In this case, instead of failing the transaction due to the error and returning the user to the checkout page, we return early from the function, since we cannot find a matching payment method, and allow the rest of the order complete page to process. This incorrectly sets the order status to `Pending payment` and suggests to the customer that the transaction has completed successfully.

This PR adjusts this flow to check for errors initially, so that we can display an error and return to the checkout screen as expected, without erroneously completing the order.




#### Testing instructions


1. Pull latest version of develop branch for this plugin, enable UPE and additional UPE payment methods through WCPay Dev Tools or by adding `_wcpay_feature_upe` to your `wp_options` table, and run the latest version of the Payment Server.
2. Add an item from your store to cart and proceed to checkout.
3. Complete the checkout details and from the UPE select either Giropay or Sofort.
4. Once you are redirected to Stripe's transaction page, select `Fail test transaction`.
5. Notice with relief and mild contentment that you have now found yourself back on the checkout page looking at an error.
6. Check the logs, with logging enabled, to find the specific error.



-------------------
